### ISSUE TITLE
InstaGAN

### ISSUE THREAD

**Issue by bsahil29** [cc-ai/kdb#67]
For training the InstaGAN model over the available amount of data we have for now, we need the images in 2 domains (flooded and non-flooded) and their segmentation masks. For the 50 annotated flood images that Sasha gave, I have extracted binary water segmentation masks but we also need the same for the other domain too i.e. non-flooded. How should I go about this @sashavor @vict0rsch ?

**Comment by sashavor** [comment]
Try using DeepLab or another pretrained model for this. It should work well
on non flooded houses.

**Comment by bsahil29** [comment]
@sashavor What is the color to label mapping that you have used for flooded images?

**Comment by vict0rsch** [comment]
Agreed, standard seg should work. We need to define the other domain though:

* what is the "floor"?
  * in terms of labels
  * pbly: sidewalk + street + grass + whatever else

**Comment by sashavor** [comment]
Categories = {'house': 1, 'water': 2, 'sky': 3, 'tree': 4, 'car' : 5,
'road' : 6, 'person': 7}

**Comment by bsahil29** [comment]
I have segmented 50 house images using the Deeplab cityscapes pretrained model. The link to same dataset is here : [House segmentation data](https://drive.google.com/open?id=1rSofnliQKdZgwLBSVblunTo-ucdS-xFy)

Link to Code : [Github Link](https://github.com/bsahil29/FloorSegmentation)

**Comment by bsahil29** [comment]
Correctly segmented 50 house images using the Deeplab cityscapes pretrained model. The link to same dataset is here : [Correct House segmentation data](https://drive.google.com/open?id=1df0AzEQvaVE4o6JQoo6G6NLD9D_jyLyb)

Removed the earlier one. There was an issue with the format used for saving the images in earlier case (changed it from jpg to png).

**Comment by bsahil29** [comment]
Managed to train the InstaGAN over 60 images (30 each from 2 domains). 
For some of the images, segmentation generated from the Deeplab cityscapes pretrained model didn't detect anything at all or the part detected was very less and I assume the InstaGAN does some kind of cropping so those kind of images were giving a bug. It finally worked for 30 images out of those 50 and to keep set size same, I only selected 30 out of 50 manually segmented flood images.

Further to Explore :
- Training with Cityscape images
- Or increase the dataset. For now, can't say anything about the performance with just 30 images
- Play with parameters, increase the batch size and observe the changes in training performance and speed

**Comment by bsahil29** [comment]
These are the some of the good results from training InstaGAN over 30 images for 600 epochs:

![Epoch 547:](https://user-images.githubusercontent.com/36884804/59058047-c3d57280-8869-11e9-81a3-d9525d3b711f.png)
![Epoch 534:](https://user-images.githubusercontent.com/36884804/59058053-c768f980-8869-11e9-834a-c416f8ac83c9.png)
![Epoch 440:](https://user-images.githubusercontent.com/36884804/59058059-ca63ea00-8869-11e9-929d-aeaadeb5251a.png)
![Epoch 427:](https://user-images.githubusercontent.com/36884804/59058063-cd5eda80-8869-11e9-846b-489fddcc19ad.png)

- The first image is the non-flooded image and the second is the corresponding flooded image. I am yet to figure out what exactly is the third image (rec_image). 
- What appears from the results is that as the segmentation for non-flooded images was not very good, hence the results suffer. If we have a better segmentation, it appears that we have a scope to use InstaGAN and we can get better results possibly.
- plus maybe 30 images weren't just enough.

**Comment by bsahil29** [comment]
Managed to integrate Cometml with InstaGAN so we can now run experiments and keep track of them. Was fairly simple.

**Comment by vict0rsch** [comment]
@bsahil29 I think the third is the REConstructed image: A -(G_a)-> Fake B -(G_b)-> Reconstructed A 

I think it's really interesting results since for the first time we can see attempts at overflooding the houses, not just inpainting :)

**Comment by bsahil29** [comment]
@vict0rsch Yeah I also think it's the reconstructed image and seeing the results I think we should improve (in terms of segmentation) and extend our dataset and see how it goes. Maybe we can attain something using InstaGAN.

**Comment by bsahil29** [comment]
@vict0rsch @sashavor We should discuss on how to go ahead with the fine tuning of InstaGAN? What we want to fine tune and what not out of the 4 nets (2 generators and 2 discriminators)?
